active/0030-emqx_linear_channel_registry.md

```diff
@@ -0,0 +1,321 @@
+# Linear Channel Registry
+
+## Changelog
+
+* 2025-04-20: @qzhuyan First draft
+
+## Abstract
+
+For a new incoming MQTT connection, the node may look for an existing channel matching the same clientid 
+within the cluster for taking further actions such as taking over the session, discarding the session, 
+or creating a new session.
+
+This channel registration info is a cluster wide global state maintained by the underlying database. 
+It is replicated asynchronously (eventually consistent), and it leads to problems when replication is 
+lagging or the old channel process is unresponsive.
+
+In this EIP, we propose another channel registration subsystem called LCR ("Linear Channel Registry") that 
+expands the channel prop with version (e.g., trpt_connected_at timestamp when transport connects) so that the 
+young and old channels could be determined to minimize the issues caused by race conditions during session takeover
+while the client is reconnecting, especially in a massive volume.
+
+## Motivation
+
+For the current implementation of channel registry (emqx_cm_registry), it has eventual consistency.
+
+The channel registration gets a dirty write on the core node from an RPC call.
+The registration then gets replicated among the core nodes asynchronously (dirty).
+The core nodes then replicate the registration to the replicant nodes.
+When looking for existing channels while handling the new connection, EMQX do a lookup from the local copy.
+
+This is a design for high availability and high performance, having the assumption that replication is finished 
+before the next client connection. But when replication starts lagging due to various reasons, the following issues pop up:
+
+- A node does not see that the channel exists. It creates a new session instead of taking over the existing one.
+
+- A new channel could take over a session from the wrong channel due to the wrong order of receiving replications.
+
+- The older channel with an unresponsive connection kicks the latest channel, causing the client to get disconnected unexpectedly.
+
+- Wrong Session Present flag in MQTT.CONNACK is returned to the client.
```

Comment:
why did this happen and how does the new algorithm fix it?

Reply:
quick answer: https://github.com/emqx/emqx/pull/15017

I will update this doc for that.